File: ffi/zig/examples/shared_memory.zig
/// Example: Zig ↔ C++ shared memory interop
///
/// Demonstrates creating a shared memory segment from Zig,
/// writing data into it, then reading it back via a second handle —
/// exactly the same pattern a C++ (or Rust, or C) process could use
/// on the other end.
const std = @import("std");
const lsm = @import("lsm");

pub fn main() !void {
    const message = "Hello from Zig!";

    // Writer: create the shared memory segment and copy a message in
    const writer = try lsm.SharedMemory.create("zigExample", 256, true);
    defer writer.deinit();

    const wbuf = writer.data();
    @memcpy(wbuf[0..message.len], message);
    std.debug.print("Wrote   : {s}\n", .{message});

    // Reader: open the same segment and read the bytes back
    const reader = try lsm.SharedMemory.open("zigExample", 256, true);
    defer reader.close();

    const received = reader.data()[0..message.len];
    std.debug.print("Received: {s}\n", .{received});

    if (!std.mem.eql(u8, received, message)) {
        return error.RoundTripMismatch;
    }
    std.debug.print("OK\n", .{});
}

File: ffi/rust/examples/shared_memory.rs
// Example: Rust ↔ C++ shared memory interop
//
// Demonstrates creating a shared memory segment from Rust,
// writing data into it, then reading it back via a second handle —
// exactly the same pattern a C++ process could use on the other end.

use libsharedmemory::SharedMemory;

fn main() {
    let segment_name = "rustExample";
    let segment_size: usize = 256;

    // Writer: create the shared memory segment and copy a message in
    let writer = SharedMemory::create(segment_name, segment_size, true)
        .expect("Failed to create shared memory segment");

    let message = b"Hello from Rust!";
    writer.as_mut_slice()[..message.len()].copy_from_slice(message);
    println!("Wrote   : {}", std::str::from_utf8(message).unwrap());

    // Reader: open the same segment and read the bytes back
    let reader = SharedMemory::open(segment_name, segment_size, true)
        .expect("Failed to open shared memory segment");

    let received = &reader.as_slice()[..message.len()];
    println!("Received: {}", std::str::from_utf8(received).unwrap());

    assert_eq!(received, message, "Round-trip mismatch!");
    println!("OK");
}
